README.md
# AceSegmentWriter

[![AUnit Tests](https://github.com/bxparks/AceSegmentWriter/actions/workflows/aunit_tests.yml/badge.svg)](https://github.com/bxparks/AceSegmentWriter/actions/workflows/aunit_tests.yml)

A companion library to the [AceSegment](https://github.com/bxparks/AceSegment)
library. This provides higher-level convenience classes for writing decimal
numbers, hex numbers, temperature, clock digits, characters, and strings to
seven segment LED modules. The following classes are provided: `PatternWriter`,
`NumberWriter`, `ClockWriter`, `TemperatureWriter`, `CharWriter`,
`StringWriter`, `LevelWriter`, and `StringScroller`.

Initially, this code was part of the AceSegment library and depended directly on
the `LedModule` class of that library. Later the direct dependency was removed
by converting the Writer classes into generic code using C++ templates. The
classes now depend on the `T_LED_MODULE` template type which should implement
the public methods that appear in the `LedModule` class. But `T_LED_MODULE` is
*not* required to inherit from `LedModule` which preserves the decoupling
between the AceSegmentWriter and AceSegment libraries.

**Version**: 0.5 (2023-03-16)

**Changelog**: [CHANGELOG.md](CHANGELOG.md)

## Table of Contents

* [Installation](#Installation)
    * [Source Code](#SourceCode)
    * [Dependencies](#Dependencies)
* [Documentation](#Documentation)
    * [Examples](#Examples)
* [High Level Overview](#HighLevelOverview)
    * [Classes and Types](#ClassesAndTypes)
    * [Dependency Diagram](#DependencyDiagram)
    * [Digit and Segment Addressing](#DigitAndSegmentAddressing)
* [Usage](#Usage)
    * [Include Header and Namespace](#HeaderAndNamespace)
    * [T_LED_MODULE](#T_LED_MODULE)
    * [PatternWriter](#PatternWriter)
    * [NumberWriter](#NumberWriter)
    * [ClockWriter](#ClockWriter)
    * [TemperatureWriter](#TemperatureWriter)
    * [CharWriter](#CharWriter)
    * [StringWriter](#StringWriter)
    * [LevelWriter](#LevelWriter)
    * [StringScroller](#StringScroller)
* [Resource Consumption](#ResourceConsumption)
    * [Flash And Static Memory](#FlashAndStaticMemory)
* [System Requirements](#SystemRequirements)
    * [Hardware](#Hardware)
    * [Tool Chain](#ToolChain)
    * [Operating System](#OperatingSystem)
* [Bugs And Limitations](#BugsAndLimitations)
* [Alternative Libraries](#AlternativeLibraries)
* [License](#License)
* [Feedback and Support](#FeedbackAndSupport)
* [Authors](#Authors)

<a name="Installation"></a>
## Installation

The latest stable release is available in the Arduino IDE Library Manager.
Search for "AceSegmentWriter". Click install.

The development version can be installed by cloning the [GitHub
repository](https://github.com/bxparks/AceSegmentWriter), checking out the
default `develop` branch, then manually copying over to or symlinking from the
`./libraries` directory used by the Arduino IDE. (The result is a directory or
link named `./libraries/AceSegmentWriter`.)

The `master` branch contains the stable releases.

<a name="SourceCode"></a>
### Source Code

The source files are organized as follows:
* `src/AceSegmentWriter.h` - main header file
* `src/ace_segment_writer/` - implementation files
* `tests/` - unit tests which require [AUnit](https://github.com/bxparks/AUnit)
* `examples/` - example sketches
* `docs/` - contains the doxygen docs and additional manual docs

<a name="Dependencies"></a>
### Dependencies

This library library has a direct, compile-time dependency on:

* AceCommon (https://github.com/bxparks/AceCommon)

The following libraries are recommended, and they are added in the `depends`
property of `library.properties` so that they are automatically installed by the
Arduino Library Manager for convenience. However, client applications are not
*required* to use these. Other libraries with the same interface and
functionality can be substituted if desired.

* AceSegment (https://github.com/bxparks/AceSegment)
* AceSPI (https://github.com/bxparks/AceSPI)
* AceTMI (https://github.com/bxparks/AceTMI)
* AceWire (https://github.com/bxparks/AceWire)

The unit tests depend on:

* AUnit (https://github.com/bxparks/AUnit)

Some of the examples may depend on the following which may need to be installed
manually:

* AceButton (https://github.com/bxparks/AceButton)

<a name="Documentation"></a>
## Documentation

* this `README.md` file
* [Doxygen docs](https://bxparks.github.io/AceSegmentWriter/html)
    * On Github pages.

<a name="Examples"></a>
### Examples

The following example sketches are provided:

* Basic
    * All of the examples assume an LED module based on the TM1637 controller
      chip. These LED modules are readily found on retail outlets like eBay or
      Amazon. You can reconfigure the examples to use a different LED module.
    * [PatternWriterDemo](examples/PatternWriterDemo)
    * [NumberWriterDemo](examples/NumberWriterDemo)
    * [ClockWriterDemo](examples/ClockWriterDemo)
    * [TemperatureWriterDemo](examples/TemperatureWriterDemo)
    * [CharWriterDemo](examples/CharWriterDemo)
    * [StringWriterDemo](examples/StringWriterDemo)
    * [LevelWriterDemo](examples/LevelWriterDemo)
    * [StringScrollerDemo](examples/StringScrollerDemo)
* Intermediate
    * [WriterTester](examples/WriterTester)
        * demo of the various `src/writer` classes
        * supports multiple LED modules based on different controller chips
          though `#define` macros
        * depends on AceButton (https://github.com/bxparks/AceButton) library
        * uses 2 buttons for "single step" debugging mode
* Benchmarks
    * [MemoryBenchmark](examples/MemoryBenchmark): determines the size of
      the various components of the library
    * No `AutoBenchmark` program.
        * These Writer classes simply write into the in-memory buffer provided
          by the underying `T_LED_MODULE` class.
        * The execution time is too fast to be easily measured and does not seem
          worth the effort.

<a name="HighLevelOverview"></a>
## High Level Overview

<a name="ClassesAndTypes"></a>
### Classes and Types

Here are the classes and types in the library:

* `T_LED_MODULE`
    * Generic type for classes in this library. It is assumed to implement the
      public methods of the `LedModule` class of the AceSegment library.
* `PatternWriter`
    * Class that knows how to write segment bit patterns to a `T_LED_MODULE`.
    * Other Writer classes are built upon this by composition.
* `NumberWriter`
    * A class that writes integers in decimal or hexadecimal format to the
      `T_LED_MODULE`.
    * A few additional characters are supported: `kDigitSpace`, `kDigitMinus`
* `ClockWriter`
    * A class that writes a clock string "hh:mm" to `T_LED_MODULE`.
    * Builds on top of `NumberWriter`.
* `TemperatureWriter`
    * A class that writes temperatures with a degrees symbol or optionally
      with "C" or "F" symbol.
    * Builds on top of `NumberWriter`.
* `CharWriter`
    * A class that convert an ASCII character represented by a `char` (code
      0-127) to a bit pattern used by `SegmentDriver` class.
    * Not all ASCII characters can be rendered on a seven segment display
      legibly but the `CharWriter` tries its best.
* `StringWriter`
    * A class that prints strings of `char` to a `CharWriter`, which in
      turns, prints to the `T_LED_MODULE`.
    * Builds on top of `CharWriter`.
* `StringScoller`
    * Scroll a string left and right.
* `LevelWriter`
    * Display specified number of bars from left to right, 2 bars per digit.

<a name="DependencyDiagram"></a>
### Dependency Diagram

The conceptual dependency diagram among these classes looks something like this:

```
StringScroller StringWriter  ClockWriter TemperatureWriter
            \       /             \      /
             v     v               v    v
             CharWriter        NumberWriter      LevelWriter
                      \             |             /
                       v            v            v
                       PatternWriter<T_LED_MODULE>
                                |
                                v
                           T_LED_MODULE
                                |
                                | (depends on AceSegment if
                                | T_LED_MODULE is set to LedModule)
                                v
                            AceSegment
                            Library
                           /    |     \
                          v     v      v
                    AceWire  AceTMI   AceSPI
                    Library  Library  Library
```

<a name="DigitAndSegmentAddressing"></a>
### Digit and Segment Addressing

The `T_LED_MODULE` type must support the same conventions for addressing
the digits and segments as defined by the AceSegment library:

* digits start at position 0 on the left and increase to the right
* segments are assigned bits 0 to 7 of an unsigned byte (type `uint8_t`) with
  segment `a` assigned to bit 0, segment `g` assigned to bit 6, and the optional
  decimal point assigned to bit 7

```
+------------+  +------------+           +------------+
|   aaaa     |  |   aaaa     |           |   aaaa     |
|  f    b    |  |  f    b    |           |  f    b    |
|  f    b    |  |  f    b    |           |  f    b    |
|   gggg     |  |   gggg     |  *  *  *  |   gggg     |
|  e    c    |  |  e    c    |           |  e    c    |
|  e    c    |  |  e    c    |           |  e    c    |
|   dddd  dp |  |   dddd  dp |           |   dddd  dp |
+------------+  +------------+           +------------+
  Digit 0         Digit 1                 Digit {N-1}


Segment: dp g f e d c b a
   Bit:  7  6 5 4 3 2 1 0
```

Some 4-digit LED modules are designed to be used in a clock to display the hour
and minute components of the time as `HH:MM`. In these modules, it is common for
the decimal point on Digit 1 to be replaced with the `colon` segment between
Digit 1 and Digit 2. In these modules, sometimes the decimal points for the
other digits work normally, but sometimes, the remaining decimal points do not
work at all.

<a name="Usage"></a>
## Usage

<a name="HeaderAndNamespace"></a>
### Include Header and Namespace

Only a single header file `AceSegmentWriter.h` is required to use this library.
To prevent name clashes with other libraries that the calling code may use, all
classes are defined in the `ace_segment` namespace. (This is the same namespace
used by the AceSegment library to make it easier to use them together.) To use
the code without prepending the `ace_segment::` prefix, use the `using`
directive:

```C++
#include <Arduino.h>
#include <AceSegmentWriter.h>
using namespace ace_segment;
```

<a name="T_LED_MODULE"></a>
### `T_LED_MODULE`

All Writer classes directly (or indirectly) wrap around an underlying LED module
class with the generic `T_LED_MODULE` type. It is assumed to implement the same
public methods of the
[LedModule](https://github.com/bxparks/AceSegment/blob/develop/src/ace_segment/LedModule.h)
class from AceSegment. However, the Writer classes in this library are
implemented as C++ templates, so the `T_LED_MODULE` type does **not** need to
inherit from the `LedModule` class. The public methods of `T_LED_MODULE`
should look like this:

```C++
class LedModule {
  public:
    uint8_t size() const;
    void setPatternAt(uint8_t pos, uint8_t pattern);
    uint8_t getPatternAt(uint8_t pos) const;
    void setBrightness(uint8_t brightness);
    void getBrightness() const;
    void setDecimalPointAt(uint8_t pos, bool state = true);
};
```

<a name="PatternWriter"></a>
### PatternWriter

The `PatternWriter` class is the most basic wrapper around an `T_LED_MODULE`
object, and provides more convenient interfaces to writing to the LED module. It
provides the following features on top of `T_LED_MODULE`:

* Validation is performed on the digit `pos` parameter. If the `pos` is not
  valid, then the method returns immediately without performing any action.
* Entire strings (both normal strings and `PROGMEM` strings) can be written
  to the led module.
* The `setDecimalPointAt()` convenience function adds a decimal point at the
  specified `pos` location.
* The `clear()` and `clearToEnd()` functions provide ways to clear the LED
  display.
* A stateful cursor that remembers its current `pos` digit in the `LedModule`.

All other `XxxWriter` classes are built on top of this class. If multiple
writers are used in an application (which happens often because each Writer is
responsible for writing different things), the application should make sure that
only a single instance of `PatternWriter` is created. Otherwise, the "current"
position may become confusing.

The public methods and constants of the class look like this:

```C++
namespace ace_segment {

const uint8_t kPattern0 = 0b00111111;
const uint8_t kPattern1 = 0b00000110;
const uint8_t kPattern2 = 0b01011011;
const uint8_t kPattern3 = 0b01001111;
const uint8_t kPattern4 = 0b01100110;
const uint8_t kPattern5 = 0b01101101;
const uint8_t kPattern6 = 0b01111101;
const uint8_t kPattern7 = 0b00000111;
const uint8_t kPattern8 = 0b01111111;
const uint8_t kPattern9 = 0b01101111;
const uint8_t kPatternA = 0b01110111;
const uint8_t kPatternB = 0b01111100;
const uint8_t kPatternC = 0b00111001;
const uint8_t kPatternD = 0b01011110;
const uint8_t kPatternE = 0b01111001;
const uint8_t kPatternF = 0b01110001;
const uint8_t kPatternSpace = 0b00000000;
const uint8_t kPatternMinus = 0b01000000;
const uint8_t kPatternDeg = 0b01100011;
const uint8_t kPatternP = 0b01110011;

template <typename T_LED_MODULE>
class PatternWriter {
  public:
    explicit PatternWriter(T_LED_MODULE& ledModule);

    T_LED_MODULE& ledModule() const;

    uint8_t size() const;
    void home();
    uint8_t pos() const;
    void pos(uint8_t pos);

    void writePattern(uint8_t pattern);
    void writePatterns(const uint8_t patterns[], uint8_t len);
    void writePatterns_P(const uint8_t patterns[], uint8_t len);
    void writeDecimalPoint(bool state = true);
    void setDecimalPointAt(uint8_t pos, bool state = true);

    void clear();
    void clearToEnd();
};

}
```

The `writePattern()` function writes the given `pattern` at the current `pos`.
The `pos` is automatically incremented by one.

The `writeDecimalPoint()` function writes a decimal point in the *previous*
digit, because that's where the decimal point is located on a seven-segment LED
display. If the current position is `0`, this function does nothing. If the
current position is equal to `ledModule.size()` which is one position after the
end, this function sets the decimal point on the last digit.

The `setDecimalPointAt()` function writes the decimal point at the given `pos`.
This provides an escape hatch in cases where `writeDecimalPoint()` is not
sufficient.

The decimal point is stored as bit 7 (the most significant bit) of the `uint8_t`
byte for a given digit. This bit is cleared by the `writePattern()` and
`writePatterns()` functions. So the `writeDecimalPoint()` and
`setDecimalPointAt() methods must be called **after** the `writePattern()`
methods.

The `pos()` function sets or gets the current position.

The `home()` function sets the current position to 0.

The `clear()` function clears all the digits of the `ledModule`. The
`clearToEnd()` clears only the digits from the current position to the end. In
both cases, the `home()` function is automatically called to set the position to
0.

Here is how to create an instance of `PatternWriter` from an instance of
`LedModule`:

```C++
PatternWriter<LedModule> patternWriter(ledModule);
```

<a name="NumberWriter"></a>
### NumberWriter

The `NumberWriter` can print integers to the `T_LED_MODULE` using decimal (0-9)
or hexadecimal (0-9A-F) formats. On platforms that support it (AVR and ESP8266),
the bit mapping table is stored in flash memory to conserve static memory.

The public methods of this class looks something like this:

```C++
namespace ace_segment {

const uint8_t kNumDigitPatterns = 18;
extern const uint8_t kDigitPatterns[kNumDigitPatterns];

typedef uint8_t digit_t;
const digit_t kDigitSpace = 0x10;
const digit_t kDigitMinus = 0x11;

template <typename T_LED_MODULE>
class NumberWriter {
  public:
    explicit NumberWriter(PatternWriter<T_LED_MODULE>& ledModule);

    T_LED_MODULE& ledModule();
    PatternWriter<T_LED_MODULE>& patternWriter();

    uint8_t size() const;
    void home();

    void writeDigit(digit_t c);
    void writeDigits(digit_t s[], uint8_t len);

    void writeDec2(uint8_t d, uint8_t padPattern = kPattern0);
    void writeDec4(uint16_t dd, uint8_t padPattern = kPattern0);

    void writeBcd(uint8_t bcd);
    void writeHexByte(uint8_t b);
    void writeHexWord(uint16_t w);

    void writeUnsignedDecimal(uint16_t num, int8_t boxSize = 0);
    void writeSignedDecimal(int16_t num, int8_t boxSize = 0);

    void writeFloat(float x, uint8_t places = 2);
    void writeDecimalPoint(bool state = true);

    void clear();
    void clearToEnd(uint8_t pos);
};

}
```

The `digit_t` type semantically represents the character set supported by this
class. It is implemented as an alias for `uint8_t`, which unfortunately means
that the C++ compiler will not warn about mixing this type with another
`uint8_t`. The range of this character set is from `[0,15]` plus 2 additional
symbols, so `[0,17]`:

* `ace_segment::kDigitSpace`
* `ace_segment::kDigitMinus`

An instance of `NumberWriter` is created like this:

```C++
PatternWriter<LedModule> patternWriter(ledModule);
NumberWriter<LedModule> numberWriter(patternWriter);
```

![NumberWriter](docs/writers/number_writer_hex.jpg)

![NumberWriter](docs/writers/number_writer_decimal.jpg)

<a name="ClockWriter"></a>
### ClockWriter

There are special, 4 digit,  seven segment LED displays which replace the
decimal point with the colon symbol ":" between the 2 digits on either side so
that it can display a time in the format "hh:mm".

The public methods of this class look like this:

```C++
namespace ace_segment {

template <typename T_LED_MODULE>
class ClockWriter {
  public:
    explicit ClockWriter(
        NumberWriter<T_LED_MODULE>& numberWriter,
        uint8_t colonDigit = 1);

    T_LED_MODULE& ledModule();
    PatternWriter<T_LED_MODULE>& patternWriter();
    NumberWriter<T_LED_MODULE>& numberWriter();

    uint8_t size() const;
    void home();

    void writeHourMinute24(uint8_t hh, uint8_t mm);
    void writeHourMinute12(uint8_t hh, uint8_t mm);
    void writeColon(bool state = true);

    void clear();
    void clearToEnd(
};

}
```

An instance of `ClockWriter` is created like this:

```C++
PatternWriter<LedModule> patternWriter(ledModule);
NumberWriter<LedModule> numberWriter(patternWriter);
ClockWriter<LedModule> clockWriter(numberWriter);
```

You can write the letters `A` and `P` using the underlying `patternWriter()`:

```
clockWriter.patternWriter().writePattern(ace_segment::kPatternA);
```

![ClockWriter](docs/writers/clock_writer.jpg)

<a name="TemperatureWriter"></a>
### TemperatureWriter

This class supports writing out temperatures in degrees Celsius or Fahrenheit.
The public methods of this class looks something like this:

```C++
namespace ace_segment {

const uint8_t kPatternDegree = 0b01100011;
const uint8_t kPatternC = 0b00111001;
const uint8_t kPatternF = 0b01110001;

template <typename T_LED_MODULE>
class TemperatureWriter {
  public:
    explicit TemperatureWriter(NumberWriter<T_LED_MODULE>& numberWriter);

    T_LED_MODULE& ledModule();
    PatternWriter<T_LED_MODULE>& patternWriter();
    NumberWriter<T_LED_MODULE>& numberWriter();

    uint8_t size() const;
    void home();

    uint8_t writeTemp(int16_t temp, boxSize = 0);
    uint8_t writeTempDeg(int16_t temp, boxSize = 0);
    uint8_t writeTempDegC(int16_t temp, boxSize = 0);
    uint8_t writeTempDegF(int16_t temp, boxSize = 0);

    void clear();
    void clearToEnd();
};

}
```

An instance of `TemperatureWriter` is created like this:

```C++
PatternWriter<LedModule> patternWriter(ledModule);
NumberWriter<LedModule> numberWriter(patternWriter);
TemperatureWriter<LedModule> temperatureWriter(numberWriter);
```

![TemperatureWriter-Celsius](docs/writers/temperature_writer_celsius.jpg)

![TemperatureWriter-Fahrenheit](docs/writers/temperature_writer_fahrenheit.jpg)

<a name="CharWriter"></a>
### CharWriter

It is possible to represent many of the ASCII characters in the range `[0,127]`
on a seven-segment LED display, although some of the characters will necessarily
be crude given the limited number of segments. The `CharWriter` contains a
[mapping of ASCII](https://github.com/dmadison/LED-Segment-ASCII) characters
to seven-segment bit patterns. On platforms that support it (AVR and
ESP8266), the bit pattern array is stored in flash memory to conserve static
memory.

The public methods of this class look like this:

```C++
namespace ace_segment {

const uint8_t kPatternUnknown = 0b00000000;

const uint8_t kNumCharPatterns = 128;
extern const uint8_t kCharPatterns[kNumCharPatterns];

template <typename T_LED_MODULE>
class CharWriter {
  public:
    explicit CharWriter(
        PatternWriter<T_LED_MODULE>& patternWriter,
        const uint8_t charPatterns[] = kCharPatterns,
        uint8_t numChars = kNumChars
    );

    T_LED_MODULE& ledModule();
    PatternWriter<T_LED_MODULE>& patternWriter();

    uint8_t getNumChars() const;
    uint8_t getPattern(char c) const;

    uint8_t size() const;
    void home();

    void writeChar(char c);

    void clear();
    void clearToEnd();
};

}
```

An instance of `CharWriter` is created like this:

```C++
PatternWriter<LedModule> patternWriter(ledModule);
CharWriter<LedModule> charWriter(patternWriter);
```

You can use a custom font by providing an array of segment bit patterns
`patterns[]` in the constructor of `CharWriter`.

![CharWriter](docs/writers/char_writer.jpg)

<a name="StringWriter"></a>
### StringWriter

A `StringWriter` is a class that builds on top of the `CharWriter`. It knows how
to write entire strings into the LED display. The public methods look like:

```C++
template <typename T_LED_MODULE>
class StringWriter {
  public:
    explicit StringWriter(CharWriter<T_LED_MODULE>& charWriter);

    T_LED_MODULE& ledModule();
    PatternWriter<T_LED_MODULE>& patternWriter();
    CharWriter<T_LED_MODULE>& charWriter();

    uint8_t size() const;
    void home();

    uint8_t writeString(const char* cs, uint8_t numChar = 255);
    uint8_t writeString(const __FlashStringHelper* fs, uint8_t numChar = 255);

    void clear();
    void clearToEnd();
};
```

The implementation of `writeString()` is straightforward except for the
handling of a decimal point. A seven segment LED digit contains a small LED for
the decimal point. Instead of taking up an entire digit for a single '.'
character, we can collapse the '.' character into the decimal point indicator of
the previous character on the left.

The optional `numChar` parameter limits the number of characters in the string
to write. The default value is 255 which is expected to be larger than the
largest LED module that will be used with the AceSegment and AceSegmentWriter
libraries, so the default value will print the entire string.

The actual number of LED digits written is returned by `writeString()`. For
example, writing `"1.2"` returns 2 because the decimal point was merged into the
previous digit and only 2 digits are written.

The `clearToEnd()` method clears the LED display from the current `pos` to the
end of the display.

An instance of `StringWriter` is created like this:

```C++
PatternWriter<LedModule> patternWriter(ledModule);
CharWriter<LedModule> charWriter(patternWriter);
StringWriter<LedModule> stringWriter(charWriter);
```

The following will write the given string and clear all digits after the end of
the string:

```
stringWriter.writeString(s);
stringWriter.clearToEnd();
```

![StringWriter](docs/writers/string_writer.jpg)

<a name="LevelWriter"></a>
### LevelWriter

A `LevelWriter` writes a specified number of vertical bars (2 vertical
bar per digit) to the LED display, emulating a level meter LED module.

```C++
namespace ace_segment {

const uint8_t kPatternLevelLeft = 0b00110000;
const uint8_t kPatternLevelRight = 0b00000110;

template <typename T_LED_MODULE>
class LevelWriter {
  public:
    explicit LevelWriter(PatternWriter<T_LED_MODULE>& patternWriter);

    T_LED_MODULE& ledModule();
    PatternWriter<T_LED_MODULE>& patternWriter();

    uint8_t getMaxLevel() const;
    void writeLevel(uint8_t level);
};

}
```

An instance of `LevelWriter` is created like this:

```C++
PatternWriter<LedModule> patternWriter(ledModule);
LevelWriter<LedModule> levelWriter(patternWriter);
```

There are 2 vertical bars available per per digit. So the maximum level
supported by a 4-digit LED module is 8, and an 8-digit LED module supports a
maximum level of 16.

![LevelWriter](docs/writers/level_writer.jpg)

<a name="StringScroller"></a>
### StringScroller

A `StringScroller` is a class that builds on top of the `CharWriter`. It can
scroll strings to the left and right. The public methods look like:

```C++
namespace ace_segment {

template <typename T_LED_MODULE>
class StringScroller {
  public:
    explicit StringScroller(CharWriter<T_LED_MODULE>& charWriter);

    T_LED_MODULE& ledModule();
    PatternWriter<T_LED_MODULE>& patternWriter();
    CharWriter<T_LED_MODULE>& charWriter();

    uint8_t size() const;

    void initScrollLeft(const char* s);
    void initScrollLeft(const __FlashStringHelper* s);
    bool scrollLeft();

    void initScrollRight(const char* s);
    void initScrollRight(const __FlashStringHelper* s);
    bool scrollRight();
};

}
```

An instance of `StringScroller` is built from its underlying classes like this:

```C++
PatternWriter<LedModule> patternWriter(ledModule);
CharWriter<LedModule> charWriter(patternWriter);
StringScroller<LedModule> stringScroller(charWriter);
```

To scroll a string to the left, initialize the string using `initScrollLeft()`,
then call `scrollLeft()` to shift one position to the left. The string scrolls
into the first digit on the right. When the scrolling is finished, the display
becomes blank, and the `scrollLeft()` method returns `true` to indicate
`isDone`.

Similarly to scroll to the right, initialize the string using
`initScrollRight()`, then call `scrollRight()` to shift to the right. The string

CHANGELOG.md
# Changelog

* Unreleased
* 0.5 (2023-03-16)
    * `PatternWriter`
        * Add `writeDecimalPoint()` which writes the decimal point in the
          *previous* digit from the current cursor.
        * Retain `setDecimalPointAt()` as an escape hatch for applications with
          advanced needs.
    * `NumberWriter`
        * Add `writeDecimalPoint()` to replace `setDecimalPointAt()`.
        * Add `writeFloat()` which delegates the floating point conversion to
          the `Print::print(float x, int digits = 2)` method that is implemented
          by the Arduino `Print` class.
        * Add `writeChar()` which supports a restricted set of ASCII characters,
          enough to implement `writeFloat()`.
            * `'0'` to `'9'`, `'-'`, and `'.'`
            * All others are printed as a space `' '`, including the space
              character itself.
* 0.4 (2023-03-15)
    * **Breaking Change**
        * `PatternWriter`
            * `PatternWriter` is now stateful and holds the current `pos`.
            * `writePattern()` writes the pattern at the current position, and
              automatically increments.
            * Various `writeXxxAt()` replaced with simpler `writeXxx()` methods.
            * Added `PatternWriter::home()`, `PatternWriter::clear()`, and
              `PatternWriter::clearToEnd()`.
        * `NumberWriter`
            * Rename `hexchar_t` to `digit_t`.
            * Rename `kHexCharXxx` constants to `kDigitXxx`.
            * Rename `writeHexCharXxxAt()` to `writeDigitXxx()`
            * Rename `writeDecXxAt()` to `writeDecXx()`
            * Rename `writeUnsignedDecimalAt()` to `writeUnsignedDecimal()`
            * Rename `writeSignedDecimalAt()` to `writeSignedDecimal()`
        * `CharWriter`
            * Rename `writeCharAt()` to `writeChar()`.
        * `TemperatureWriter`
            * Rename `writeXxxAt()` to `writeXxx()`
* 0.3 (2022-02-02)
    * Upgrade to AceTMI v0.5.
        * Rename `SimpleTmiInterface` to `SimpleTmi1637Interface`.
        * Rename `SimpleTmiFastInterface` to `SimpleTmi1637FastInterface`.
    * Upgrade tool chain and regenerate `MemoryBenchmark` and `AutoBenchmark`.
        * Arduino IDE from 1.8.13 to 1.8.19
        * Arduino CLI from 0.14.0 to 0.19.2
        * Arduino AVR Core from 1.8.3 to 1.8.4
        * STM32duino from 2.0.0 to 2.2.0
        * ESP8266 Core from 2.7.4 to 3.0.2
        * ESP32 Core from 1.0.6 to 2.0.2
        * Teensyduino from 1.53 to 1.56
    * Downgrade SAMD21 boards to new "Tier 3: May work but unsupported" tier.
* 0.2 (2021-08-17)
    * Consolidate commonly used `kPatternXxx` constants into `PatternWriter.h`
      instead of being distributed across different classes.
    * Replace `writeHourMinute()` into `writeHourMinute12()` (12-hour format
      with prepending space), and `writeHourMinute24()` (24-hour format with
      prepending 0).
    * Move `writeBcd2At()`, `writeDec2At()` and `writeDec4At()` from
      `ClockWriter` to `NumberWriter`.
    * Add `padPattern` parameter to `writeDec2At()` and `writeDec4At()` with
      the default of `kPattern0` to get leading 0 characters.
    * Remove `writeUnsignedDecimal2At()` since it has the same functionality as
      `writeDec2At()`.
    * First public release.
* 0.1 (2021-08-10)
    * Initial release.
* (2021-08-09)
    * Initial extraction of [AceSegment](https://github.com/bxparks/AceSegment).


src/ace_segment_writer/CharWriter.cpp
/*
MIT License

Copyright (c) 2018 Brian T. Park

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include <Arduino.h>
#include "CharWriter.h"

namespace ace_segment {

// Bit patterns for ASCII characters (0 - 127).
// Adapted from https://github.com/dmadison/LED-Segment-ASCII.
//
// 7-segment map:
//       AAA       000
//      F   B     5   1
//      F   B     5   1
//       GGG       666
//      E   C     4   2
//      E   C     4   2
//       DDD  DP   333  77
//
// Segment: DP G F E D C B A
//    Bits: 7  6 5 4 3 2 1 0
//
const uint8_t kCharPatterns[] PROGMEM = {
  kPatternUnknown, /* 00 */
  kPatternUnknown, /* 01 */
  kPatternUnknown, /* 02 */
  kPatternUnknown, /* 03 */
  kPatternUnknown, /* 04 */
  kPatternUnknown, /* 05 */
  kPatternUnknown, /* 06 */
  kPatternUnknown, /* 07 */
  kPatternUnknown, /* 08 */
  kPatternUnknown, /* 09 */
  kPatternUnknown, /* 10 */
  kPatternUnknown, /* 11 */
  kPatternUnknown, /* 12 */
  kPatternUnknown, /* 13 */
  kPatternUnknown, /* 14 */
  kPatternUnknown, /* 15 */
  kPatternUnknown, /* 16 */
  kPatternUnknown, /* 17 */
  kPatternUnknown, /* 18 */
  kPatternUnknown, /* 19 */
  kPatternUnknown, /* 20 */
  kPatternUnknown, /* 21 */
  kPatternUnknown, /* 22 */
  kPatternUnknown, /* 23 */
  kPatternUnknown, /* 24 */
  kPatternUnknown, /* 25 */
  kPatternUnknown, /* 26 */
  kPatternUnknown, /* 27 */
  kPatternUnknown, /* 28 */
  kPatternUnknown, /* 29 */
  kPatternUnknown, /* 30 */
  kPatternUnknown, /* 31 */
  0b00000000, /* (space) */
  0b10000110, /* ! */
  0b00100010, /* " */
  0b01111110, /* # */
  0b01101101, /* $ */
  0b11010010, /* % */
  0b01000110, /* & */
  0b00100000, /* ' */
  0b00101001, /* ( */
  0b00001011, /* ) */
  0b00100001, /* * */
  0b01110000, /* + */
  0b00010000, /* , */
  0b01000000, /* - */
  0b10000000, /* . */
  0b01010010, /* / */
  0b00111111, /* 0 */
  0b00000110, /* 1 */
  0b01011011, /* 2 */
  0b01001111, /* 3 */
  0b01100110, /* 4 */
  0b01101101, /* 5 */
  0b01111101, /* 6 */
  0b00000111, /* 7 */
  0b01111111, /* 8 */
  0b01101111, /* 9 */
  0b00001001, /* : */
  0b00001101, /* ; */
  0b01100001, /* < */
  0b01001000, /* = */
  0b01000011, /* > */
  0b11010011, /* ? */
  0b01011111, /* @ */
  0b01110111, /* A */
  0b01111100, /* B */
  0b00111001, /* C */
  0b01011110, /* D */
  0b01111001, /* E */
  0b01110001, /* F */
  0b00111101, /* G */
  0b01110110, /* H */
  0b00110000, /* I */
  0b00011110, /* J */
  0b01110101, /* K */
  0b00111000, /* L */
  0b00010101, /* M */
  0b00110111, /* N */
  0b00111111, /* O */
  0b01110011, /* P */
  0b01101011, /* Q */
  0b00110011, /* R */
  0b01101101, /* S */
  0b01111000, /* T */
  0b00111110, /* U */
  0b00111110, /* V */
  0b00101010, /* W */
  0b01110110, /* X */
  0b01101110, /* Y */
  0b01011011, /* Z */
  0b00111001, /* [ */
  0b01100100, /* \ */
  0b00001111, /* ] */
  0b00100011, /* ^ */
  0b00001000, /* _ */
  0b00000010, /* ` */
  0b01011111, /* a */
  0b01111100, /* b */
  0b01011000, /* c */
  0b01011110, /* d */
  0b01111011, /* e */
  0b01110001, /* f */
  0b01101111, /* g */
  0b01110100, /* h */
  0b00010000, /* i */
  0b00001100, /* j */
  0b01110101, /* k */
  0b00110000, /* l */
  0b00010100, /* m */
  0b01010100, /* n */
  0b01011100, /* o */
  0b01110011, /* p */
  0b01100111, /* q */
  0b01010000, /* r */
  0b01101101, /* s */
  0b01111000, /* t */
  0b00011100, /* u */
  0b00011100, /* v */
  0b00010100, /* w */
  0b01110110, /* x */
  0b01101110, /* y */
  0b01011011, /* z */
  0b01000110, /* { */
  0b00110000, /* | */
  0b01110000, /* } */
  0b00000001, /* ~ */
  kPatternUnknown, /* (del) */
};

}
